### PR TITLE
Python app/processor

### PR DESCRIPTION
**Pull request by petersteneteg**
## Python Processors
Inviwo Processors can now be implemented directly in Python by creating a python class and deriving from inviwopy.Processor.
Bellow follows an example of a python processor:
```py
# Name: PythonExample 

import inviwopy as ivw

class PythonExample(ivw.Processor):
    def __init__(self, id, name):
        ivw.Processor.__init__(self, id, name)
        self.inport = ivw.data.VolumeInport("inport")
        self.addInport(self.inport)
        self.outport = ivw.data.VolumeOutport("outport")
        self.addOutport(self.outport)

        self.slider = ivw.properties.IntProperty("slider", "slider", 0, 0, 100, 1)
        self.addProperty(self.slider)

    @staticmethod
    def processorInfo():
        return ivw.ProcessorInfo(
            classIdentifier = "org.inviwo.PythonExample",
            displayName = "Python Example", 
            category = "Python",
            codeState = ivw.CodeState.Stable,
            tags = ivw.Tags.PY
        )

    def getProcessorInfo(self):
        return PythonExample.processorInfo()

    def initializeResources(self):
        print("init")

    def process(self):
        print("process: ", self.slider.value)
        self.outport.setData(self.inport.getData())

```
The initial '# Name:' comment is needed for Inviwo to know what python class that is should look for.
To register an Inviwo python processor one can put in in the <user setttings folder>/python_processor or by adding a `PythonProcessorFolderObserver` to an Inviwo module and have that observe a folder.

## Python Applications
It is now possible to run inviwo directly from python using a new inviwopyapp python package found in `apps/inviwopyapp`.
An example of running inviwo can be found in `apps/inviwopyapp/inviwo.py`. 
```py
    import inviwopy as ivw
    import inviwopyapp as qt
    
    if __name__ == '__main__':
        # Inviwo requires that a logcentral is created.
        lc = ivw.LogCentral()
        
        # Create and register a console logger
        cl = ivw.ConsoleLogger()
        lc.registerLogger(cl)
    
        # Create the inviwo application
        app = qt.InviwoApplicationQt()
        app.registerModules()
    
        # load a workspace
        app.network.load(app.getPath(ivw.PathType.Workspaces) + "/boron.inv")
    
        # Make sure the app is ready
        app.update()
        app.waitForPool()
        app.update()
        # Save a snapshot
        app.network.Canvas.snapshot("snapshot.png") 
    
        # run the app event loop
        app.run()
```

### CODE REVIEW

File: modules/python3/bindings/src/pyinviwoapplication.cpp [hunk 1]
```diff
@@ -31,11 +31,26 @@
...
 #include <inviwopy/vectoridentifierwrapper.h>
 
 namespace inviwo {
 
+class InviwoApplicationTrampoline : public InviwoApplication {
+public:
+    using InviwoApplication::InviwoApplication;
+
+    virtual void closeInviwoApplication() override {
+        PYBIND11_OVERLOAD(void, InviwoApplication, closeInviwoApplication, );
```
Comment (r-englund) [inline modules/python3/bindings/src/pyinviwoapplication.cpp:44]: comma -nothing? is this valid code?

File: modules/python3/bindings/src/pyinviwomodule.cpp [hunk 2]
```diff
@@ -51,11 +121,71 @@ void exposeInviwoModule(pybind11::module &m) {
...
         .def_property_readonly("description", &InviwoModule::getDescription)
         .def_property_readonly("path", [](InviwoModule *m) { return m->getPath(); })
         .def_property_readonly("version", &InviwoModule::getVersion)
-        .def("getPath", [](InviwoModule *m, ModulePath type) { return m->getPath(type); });
+        .def("getPath", [](InviwoModule *m, ModulePath type) { return m->getPath(type); })
+        .def("registerProcessor", [](InviwoModule *m, py::object pfo) {
+            m->registerProcessor(std::make_unique<ProcessorFactoryObjectPythonWrapper>(pfo));
+        });
+    //.def("registerDataReader", &InviwoModule::registerDataReader)
+    //.def("registerDataWriter", &InviwoModule::registerDataWriter);
```
Comment (r-englund) [inline modules/python3/bindings/src/pyinviwomodule.cpp]: remove commented code

File: modules/python3/bindings/src/pyinviwomodule.cpp [hunk 3]
```diff
@@ -51,11 +121,71 @@ void exposeInviwoModule(pybind11::module &m) {
...
+    .def("registerInport", &InviwoModule::registerInport)
+    .def("registerOutport", &InviwoModule::registerOutport)
+    .def("registerProperty",
+         static_cast<void (InviwoModule::*)(std::unique_ptr<PropertyFactoryObject>)>(
+             &InviwoModule::registerProperty))
+    .def("registerPropertyWidget",
+         static_cast<void (InviwoModule::*)(std::unique_ptr<PropertyWidgetFactoryObject>)>(
+             &InviwoModule::registerPropertyWidget))
+    .def("registerSettings", &InviwoModule::registerSettings);
+    */
```
Comment (r-englund) [inline modules/python3/bindings/src/pyinviwomodule.cpp:189]: remove commented code

File: modules/python3/bindings/src/pynetwork.cpp [hunk 4]
```diff
@@ -38,10 +38,14 @@
 #include <inviwo/core/ports/port.h>
 #include <inviwo/core/ports/inport.h>
 
+#include <inviwopy/vectoridentifierwrapper.h>
+
 namespace py = pybind11;
 
 namespace inviwo {
 
+using ProcessorVecWrapper = VectorIdentifierWrapper<std::vector<Processor*>>;
```
Comment (r-englund) [inline modules/python3/bindings/src/pynetwork.cpp]: Is this used anywhere?  (Searching for it in this PR gives only this match)

File: modules/python3/bindings/src/pyinviwoapplication.cpp (hunk 1 above)
Comment (r-englund) [reply modules/python3/bindings/src/pyinviwoapplication.cpp:44]: I see it in more places now, guess it some C PreProcessor thing that allows this.

File: modules/python3/bindings/src/pyinviwomodule.cpp [hunk 5]
```diff
@@ -51,11 +121,71 @@ void exposeInviwoModule(pybind11::module &m) {
...
         .def_property_readonly("identifier", &InviwoModule::getIdentifier)
         .def_property_readonly("description", &InviwoModule::getDescription)
         .def_property_readonly("path", [](InviwoModule *m) { return m->getPath(); })
         .def_property_readonly("version", &InviwoModule::getVersion)
-        .def("getPath", [](InviwoModule *m, ModulePath type) { return m->getPath(type); });
+        .def("getPath", [](InviwoModule *m, ModulePath type) { return m->getPath(type); })
+        .def("registerProcessor", [](InviwoModule *m, py::object pfo) {
+            m->registerProcessor(std::make_unique<ProcessorFactoryObjectPythonWrapper>(pfo));
+        });
+    //.def("registerDataReader", &InviwoModule::registerDataReader)
```
Comment (ResearchDaniel) [inline modules/python3/bindings/src/pyinviwomodule.cpp]: Remove?

File: modules/python3/bindings/src/pyinviwomodule.cpp [hunk 6]
```diff
@@ -51,11 +121,71 @@ void exposeInviwoModule(pybind11::module &m) {
...
+        .def_readonly("name", &InviwoModuleFactoryObject::name)
+        .def_readonly("version", &InviwoModuleFactoryObject::version)
+        .def_readonly("description", &InviwoModuleFactoryObject::description)
+        .def_readonly("inviwoCoreVersion", &InviwoModuleFactoryObject::inviwoCoreVersion)
+        .def_readonly("dependencies", &InviwoModuleFactoryObject::dependencies)
+        .def_readonly("aliases", &InviwoModuleFactoryObject::aliases)
+        .def_readonly("licenses", &InviwoModuleFactoryObject::licenses)
+        .def_readonly("protectedModule", &InviwoModuleFactoryObject::protectedModule);
+
+    /*
```
Comment (ResearchDaniel) [inline modules/python3/bindings/src/pyinviwomodule.cpp]: remove?

File: modules/python3/src/processors/pythonscriptprocessor.cpp [hunk 7]
```diff
@@ -39,8 +39,8 @@ const ProcessorInfo PythonScriptProcessor::processorInfo_{
     "org.inviwo.PythonScriptProcessor",  // Class identifier
     "Python Script Processor",           // Display name
     "Python",                            // Category
-    CodeState::Experimental,             // Code state
-    "Python, CPU",                       // Tags
+    CodeState::Deprecated,               // Code state
+    "PY",                                // Tags
```
Comment (ResearchDaniel) [inline modules/python3/src/processors/pythonscriptprocessor.cpp]: Tags::PY?

File: src/qt/applicationbase/inviwoapplicationqt.cpp [hunk 8]
```diff
@@ -55,12 +55,27 @@
 
 namespace inviwo {
 
+namespace {
+
+int dummyArgc = 1;
+
+char** dummyArgs() {
```
Comment (ResearchDaniel) [inline src/qt/applicationbase/inviwoapplicationqt.cpp:62]: Should be a comment about why this is here if

**Review by ResearchDaniel — APPROVED** [review]
Looks very promising! Only some minor comments

File: modules/python3/processors/PythonExample.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# Name: PythonExample
```
Comment (r-englund) [inline modules/python3/processors/PythonExample.py:1]: File name is using camel case. Does it have to to match with the class name?

File: modules/python3/src/processors/pythonscriptprocessor.cpp (hunk 7 above)
Comment (r-englund) [inline modules/python3/src/processors/pythonscriptprocessor.cpp]: Use `Tag::PY`

**Review by r-englund — APPROVED** [review]
nice work, just a few minor comments

**Review by r-englund — CHANGES_REQUESTED** [review]
This should have regression test!

File: modules/python3/bindings/src/pyinviwoapplication.cpp (hunk 1 above)
Comment (petersteneteg) [reply modules/python3/bindings/src/pyinviwoapplication.cpp:44]: Has to be like that, some macro va args crap.

File: modules/python3/processors/PythonExample.py (hunk 9 above)
Comment (petersteneteg) [reply modules/python3/processors/PythonExample.py:1]: As long as you have the name comment the filename can be anything.py.. if there is no name comment it uses the filename

**Review by r-englund — APPROVED** [review]

**Pull request merged**